libs/SalesforceSDK/src/com/salesforce/androidsdk/security/SalesforceKeyGenerator.java

```diff
@@ -135,6 +154,47 @@ public static String getSHA256Hash(String privateKey) {
         return hashedString;
     }
 
+    /**
+     * Generate a keypair and get the encoded public key string.
+     *
+     * @param name Alias of the entry in which the generated key will appear in Android KeyStore.
+     * @return RSA public key string.
+     */
+    public static synchronized String getRSAPublicString(String name, int length) {
+        String publicKeyBase64 = null;
+        createRSAKeys(name, length);
+        try {
+            publicKeyBase64 = Base64.encodeToString(loadKeyStore().getCertificate(name).getPublicKey().getEncoded(),
+                    Base64.NO_WRAP | Base64.NO_PADDING | Base64.URL_SAFE);
+
+        } catch (Exception e) {
+            SalesforceSDKLogger.e(TAG, "Security exception thrown", e);
+        }
+        return publicKeyBase64;
+    }
+
+    /**
+     * Get the private key from the generated key pair.
+     *
+     * @param name Alias of the entry in which the generated key will appear in Android KeyStore.
+     * @return A private key for decryption.
+     */
+    public static synchronized PrivateKey getRSAPrivateKey(String name, int length) {
+        PrivateKey privateKey = null;
+        createRSAKeys(name, length);
+        try {
+            KeyStore.Entry entry = loadKeyStore().getEntry(name, null);
```

Comment:
Shouldn't we try to call `getEntry(name, null)` first before we call `createRSAKeys`? If we have already created the keys they will exist in the `KeyStore`. However, if that call returns `null` or throws an `Exception`, then we should call `createRSAKeys`. If `createRSAKeys` returns `KeyPair` that would plug in here directly in this case.